New Intercept Weapon Special

hladkyjakub opened the pull request:
New weapon special for future use in the Dunefolk faction and very likely in the UtBS campaign.

Weapon special makes the unit deliver one strike to enemies moving adjacent, without retaliation and battle XP, but kill XP is given.

A good idea would be to make a proper API to make more abilities, which do launch extra (and non-default ruled combat) attacks (I have made quite some) fully possible.
Draw_weapon and sheath_weapon animations aren't perfect, because there aren't tools for Lua animator to make them smooth, as well as units sometimes "blicking" to the default stance between animations (precauculating and putting all animations at one spot doesn't help that)

https://github.com/user-attachments/assets/425d17d1-d6ef-462d-943a-836e0977a8e4

 #10372

hladkyjakub committed:
Added weapon special intercept

ProditorMagnus commented:
This has slightly different mechanics compared to previous PR. That is why it should be first agreed what the ability should do. Not much point to review implementation details if we dont know what code is supposed to achieve.

hladkyjakub commented:
> This has slightly different mechanics compared to previous PR. That is why it should be first agreed what the ability should do. Not much point to review implementation details if we dont know what code is supposed to achieve.

The only diference is, that it defines Weapon directly and that it doesn't have limit to once per turn (simple to change).

rodolphecombe commented:
A few comments.
I would change [attack] to [intercept] or whatever the name of the ability will be.
Line 290 to remove.

hladkyjakub commented:
> I was looking for detailed requirements like

How it works now:


Given

    * first unit has weapon special

    * first unit is not petrified

    * I think it's better off without limit, (it deals only damage equal to weapon damage)


When

    * second unit of non-allied side /enters adjacency of first unit

    * no exeptions right now, maybe skirmisher shoul be unaffected

Then

	* unit attacks second unit, like in a regular one round combat with attacker having one strike and second unit having no weapon

    * Only exeption is: that no combat XP is given (only XP for Kill)
    

@Hejnewar

Hejnewar commented:
How it should work:

Given

* first unit has weapon special
* first unit is not petrified

* **first unit has not used ability since its/global last turn end**

**There is very good reason to have it be only once per turn instead unlimited number of times. And that reason is giving players the ability to play around it. If it is unlimited, you are just subjected to the effect of this ability no matter what you do up until the unit is killed. Yet when it is limited, player can on purpose disable it and the power of this ability is also reduced making it easier fit for current use case.**  

When

* second unit of non-allied side /enters adjacency of first unit
* no exeptions right now, maybe skirmisher shoul be unaffected

**I have no problem with skirms being affected, it is simpler with every unit being affected and without exceptions.** 

Then

* unit attacks second unit, like in a regular one round combat with attacker having **number of strikes equal to weapon number of strikes** and second unit having no weapon

**I dont really have strong arguments here except for personal preference and that imo it better fits the current use case as it scales better with levels.** 

* Only exeption is: that no combat XP is given (only XP for Kill)

**One bigger comment I have is that it might be good to consinder making this a weapon special tag with few keys allowing for customisation, I always was a big fan of customisation.**

hladkyjakub commented:
Not sure if an action is expected from me at this point...
But from previous experiences, I won't take any further actions until I'm assured that this is going to get in-game (after suggested reworks)

Hejnewar commented:
I also have a question, what is expected of me right now. Mostly form organization side.

Hejnewar commented:
Okay. 
I have approval of:
SP lead - Dalas
MP lead - me
Balance team - vote passed without veto
Mainline content lead and Project Manager - Pentarctagon - on the condition of visual clarity of the skill 
("I think it does need to be made visually obvious to players when just moving to a tile is dangerous
like a different color on a unit's movement range highlight or something")

Other approvals ralate to the state of the code in PR itself like General code (ocalot) or WML / WFL / lua (Celtic_Minstrel).

Board approval is not necessary as it is never mentioned that they have to oversee certain areas of development instead it falls under project manager, specific area leads (designated by project manager) and 3 main core developers (General Code, Mainline Content, Art Direction). 

Possibly could be nice to get some help / review on the UI side from vultraz however thats also later stage thing.

Anything else?

ProditorMagnus commented:
> ("I think it does need to be made visually obvious to players when just moving to a tile is dangerous
like a different color on a unit's movement range highlight or something")

That is the next goal for developers then.

Hejnewar commented:
<img width="754" height="714" alt="Zrzut ekranu 2025-07-27 172352" src="https://github.com/user-attachments/assets/c8b8593e-db3e-4337-a892-bb03890f8b8c" />

Today I made a mockup which would look really good but is probably too hard to do for a right now single ability unless there actually is a simplish way to do this. Anyway I will leave it here just in case.

hladkyjakub commented:
I kinda hoped  in more discusion about the implementation _before_ I'll do the work and implement it, so it doesn't have to be changed later, but okay.
...
What changes do you want to be made?

Hejnewar commented:
Since the concept above lacks API to my knowledge and wont likely get one soon, how about something like this 
<img width="400" height="400" alt="damage-area" src="https://github.com/user-attachments/assets/cf5a632a-7dda-4431-8516-c950c502d196" />
Appearing around skirmisher when (enemy) unit is selected?

rodolphecombe commented:
Is it even possible to have it to appear only when an enemy unit is selected?
Otherwise, having the halo always would work.

hladkyjakub commented:
> Is it even possible to have it to appear only when an enemy unit is selected?
> Otherwise, having the halo always would work. But in that case, I would go for something more discreet, like a 144x144 red hexagon

It is possible, even in Lua, but I don't have time to look into how it's done.
However anyone is welcomed to do it.

ProditorMagnus commented:
It is possible but completely useless, so no point to try it.

Hejnewar commented:
```
    [event] 
        [filter]
        #this probably should be left empty
        [/filter]
        name=select
        first_time_only =no
        [lua]
            code=<<local halo_id = "items/spear-javelin.png"
            local current_halo_locations = {}

            local function has_active_ability(unit)
                for _, attack in ipairs(unit.attacks) do
                    for special_tag in wml.child_range(attack.specials,"attack") do
                        if special_tag.id == "intercept" then
                            return true
                        end
                    end
                end
                return false
            end

            wesnoth.game_events.on_mouse_move = function()
                -- remove previous halo
                for _, loc in ipairs(current_halo_locations) do
                    wesnoth.interface.remove_hex_overlay(loc[1], loc[2], halo_id)
                end
                current_halo_locations = {}
                local x, y = wesnoth.interface.get_selected_hex()
                if x then
                    local current_unit = wesnoth.units.get(x, y)

                    if current_unit then
                        local hexes_in_range = wesnoth.paths.find_reach(current_unit)

                        for _, candidate_hex in ipairs(hexes_in_range) do
                            local n, ne, se, s, sw, nw = wesnoth.map.get_adjacent_hexes(candidate_hex)
                            for _, adjacent_hex in ipairs({n, ne, se, s, sw, nw}) do
                                local cx = adjacent_hex[1]
                                local cy = adjacent_hex[2]
                                local enemy_unit = wesnoth.units.get(cx, cy)
                                if enemy_unit and wesnoth.sides.is_enemy(enemy_unit.side, wesnoth.current.side) and has_active_ability(enemy_unit) then
                                    wesnoth.interface.add_hex_overlay(candidate_hex[1], candidate_hex[2], { image = halo_id })
                                    table.insert(current_halo_locations, {candidate_hex[1], candidate_hex[2]}) 
                                end
                            end
                        end
                        -- both redraws are needed first one to instantly show up and the other to instantly dissapear 
                        wml.fire("redraw")
                    end
                end
                wml.fire("redraw")
            end
            --iterate throught tiles with BFS, (so it's not throught whole map, there is function that does that already)
            --if tile is adjacent to unit with the special, do mark it with the halo
            >>
        [/lua]
    [/event]
```

Here is the code that solves the visual clarity problem. The halo used can be decided by someone else. Thanks to Jarom (big help) and ZK for help with it. Currently there is no on_unit_seleced so unfortunatley it will only appear after mouse move and not on select.

ProditorMagnus commented:
You should not break other addons that rely on mouse events.

Hejnewar commented:
Can you explain please? Im not at the level of understanding without explanation.

hladkyjakub commented:
> Can you explain please? Im not at the level of understanding without explanation. 

You have to hook the function, so it does what it did + what we want it to do.

Hejnewar commented:
I tried to read the API, I tried to do something and ask. Game froze. How do you do that?

hladkyjakub commented:
> I tried to read the API, I tried to do something and ask. Game froze. How do you do that? 

Let's say that you want to hook function a() and make it do a() and b()
You do ot like this:
a_old = a
function a()
a_old()
b()
end

hladkyjakub committed:
Rename attack to extra_attack in weapon_specials.cfg

hladkyjakub commented:
@Hejnewar  done what asked.

hladkyjakub commented:
The commit name auto generated, any idea how to change it's name?

hladkyjakub commented:
There is one miss the "id= started_invisible" is leftover was in time press and can't edit it on phone, but will be corrected.

hladkyjakub committed:
Behaviour of once-per turn behaviour set just as required

hladkyjakub committed:
Implement intercept ability hex overlay logic

hladkyjakub commented:
Added Hejnes overlay code :D (with minor adjustment to not highlight hexes around units who do have the weapon special, but have used it this turn already)

I suppose the code is ready to review.

gfgtdf commented:
whats the state of this? Is there agreement that this is a feature we want?

hladkyjakub committed:
Rename game event triggers for hits and misses

hladkyjakub committed:
Refactor animator variable declarations

hladkyjakub committed:
whitespace fix

hladkyjakub committed:
update (given up on the commit names, will be squahed anyways... squashed - what a nice, squishy word :P)

hladkyjakub committed:
More flexible ability specifications

hladkyjakub committed:
last stand of whitespace (hopefully)

Updated descriptions and special notes for the intercept ability.

soliton- commented:
> Is there a way to merge commits such as whitespace-fix1, whitespace-fix2 ... into one? (ofc I would merge the whitespace with the relevant ones, but that is just an example)

From the github website? No. With a proper git client you can do whatever you want.

hladkyjakub committed:
Had to take into account that number is something else in different contexts -_-

hladkyjakub committed:
graphic logic fix

hladkyjakub committed:
bugfix (this fix is really weird but was reported as bug solving)

gfgtdf commented on data/core/macros/weapon_specials.cfg:
```diff
@@ -254,3 +254,395 @@
+local sheath_weapon = wesnoth.units.create_animator()
+
+for iter,e in ipairs(enemies) do
+    e.variables.intercept_used = e.variables.intercept_used and e.variables.intercept_used + 1 or 1
```
imo a ` e.variables.intercept_used = (e.variables.intercept_used or 0) + 1` is clearer

rodolphecombe commented:
> Some more recent showcases and possible images for this ability.
> Nagrywanie.ekranu.2025-10-01.185910.mp4
> Nagrywanie.ekranu.2025-10-01.190447.mp4

I personally prefer the first one. I am not fond of the icon in the second one.

mentos987 commented on data/core/macros/weapon_specials.cfg:
```diff
@@ -254,3 +254,358 @@
+local _ = wesnoth.textdomain "wesnoth-Bandits_from_Brown_Hills"
+local x1 = wesnoth.current.event_context.x1
+local y1 = wesnoth.current.event_context.y1
+local u = wesnoth.units.get(x1,y1)
```
Ye, it seems like it could be fun. And it seems realistic, as if though the unit attempted to intercept the moving opponent but got blocked by others while doing so.

Just make sure that the u = wesnoth.units.get(x1,y1) is hostile to the attacker.

gfgtdf commented:
> Some more recent showcases and possible images for this ability.
> Nagrywanie.ekranu.2025-10-01.185910.mp4
> Nagrywanie.ekranu.2025-10-01.190447.mp4

i also prefer the first one

gfgtdf commented:
> This ability is basically done. But due to one animation bug it is not ready to be commited.
> 
> The bug is as follows. When killing a unit sometimes the images used by this ability does not get cleaned up as if variable responsible for them is overwritten while thats going on which could happen but at the same time is very unlikely and I dont know the graphical pipeline well enough to actually say what is going on. While the solution could be global clear of an image exclusive to this ability it would also clean up any of that image that anyone would use in thier works which is imperfect but a solution.

hmm, sounds strange, do you know whether `clear_halo_locations` just doesn't get called or is it that `clear_halo_locations` faisl to clear the images?

Hejnewar commented:
> hmm, sounds strange, do you know whether clear_halo_locations just doesn't get called or is it that clear_halo_locations faisl to clear the images?

It doesnt actually happen every time and it requires good timing for me to catch it actually.

Even if it doesn't get called at that moment, we expect it to clear the images on the next pass, since until the function finishes its work, it should not clear the table contents.

But it is not what we observe.

Here are the videos. However I would wait until [hladkyjakub](https://github.com/hladkyjakub) updates the code to the most recent version. 

https://github.com/user-attachments/assets/3b1b82f7-e617-4f7d-8250-4580b726aaa8


https://github.com/user-attachments/assets/61604c14-a45a-4e75-93be-8ffc7711721e

It actually lacks one video of a blue unit not being there anymore in which case it would not get cleared at all. (Clear in this case clears whole stack of markers.)

mentos987 commented:
Ah, I see. It seems like a simple thing. Thank you for the video.

Most likely there is a list of tiles that get invalidated/queued for redraw when the unit is deselected. But the unit dying cause the chain of events to skip that part. We just need to make sure the invalidation happens in all cases.

I'll ask emperor if he want to do it himself or want my input.

hladkyjakub committed:
Refactor intercept special logic and improve clarity - bug present

hladkyjakub committed:
no change, just running CI

first time it seems github had problems on it's side

hladkyjakub committed:
whitespace fix

hladkyjakub committed:
wht

hladkyjakub committed:
Merge branch 'wesnoth:master' into Weapon-Special-Intercept

hladkyjakub committed:
Remove duplicit MACRO definition from UTBS

hladkyjakub committed:
cleaned, fixed little graphic and function bug

hladkyjakub committed:
turns out comments also have to be indented

Hejnewar reviewed (approved):
I tried but I couldnt get the UI part to break. That is good enough for me.

hladkyjakub committed:
celtic's var naming fix + goto remove

hladkyjakub committed:
more minor formating + making vars local

hladkyjakub committed:
Update weapon_specials.cfg

Removed redundant assignment of 'attacking_weapon'.

hladkyjakub committed:
Merge branch 'wesnoth:master' into Weapon-Special-Intercept

hladkyjakub committed:
Add intercept.lua

hladkyjakub committed:
Add intercept graphics

hladkyjakub committed:
moved - god only knows if this works

hladkyjakub committed:
Update intercept_graphics.lua

Remove check for double loading of intercept graphics.

hladkyjakub committed:
fix

hladkyjakub committed:
fix

Removed redundant variable assignment for 'attacking_weapon'.

hladkyjakub committed:
Update intercept_graphics.lua

hladkyjakub committed:
Update intercept logic to handle special conditions

hladkyjakub committed:
added arguments to the special

hladkyjakub committed:
Merge branch 'wesnoth:master' into Weapon-Special-Intercept

hladkyjakub committed:
Update bolas special to limit uses per turn

CelticMinstrel commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+end
+local u = wesnoth.units.get(x1,y1)
+
+local enemies = wesnoth.units.find_on_map({{"filter_adjacent",{id=u.id, is_enemy = true}},{"has_attack",{special_id = "intercept"}}})
```
This needs to use `wml.tag`.

```suggestion
local enemies = wesnoth.units.find_on_map{wml.tag.filter_adjacent{id=u.id, is_enemy = true}},wml.tag.has_attack{special_id = "intercept"}
```

CelticMinstrel commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+            wesnoth.game_events.fire("attacker_hits", e.loc, u.loc, {{"first",w.__cfg},{"second",w2_cfg},damage_inflicted=damage_inflicted})
+                if u.valid == nil or u.hitpoints <= 0 then
+                sheath_weapon:run()
+                wesnoth.game_events.fire("attack end", u.loc, loc2, {{"first",wml.get_child(wesnoth.current.event_context,"weapon")},{"second",wml.get_child(wesnoth.current.event_context,"second_weapon")}})
```
This needs to use `wml.tag`.

```suggestion
                wesnoth.game_events.fire("attack end", u.loc, loc2, {wml.tag.first(wml.get_child(wesnoth.current.event_context,"weapon")),wml.tag.second(wml.get_child(wesnoth.current.event_context,"second_weapon"))})
```

CelticMinstrel commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+            animator:add(u, "defend", "miss",{with_bars = true, primary = w2, secondary = w})--inverted
+            animator:run()
+            animator:clear()
+            wesnoth.game_events.fire("attacker_misses", e.loc, u.loc, {{"first",w.__cfg},{"second",w2_cfg}})
```
This needs to use `wml.tag`.

```suggestion
            wesnoth.game_events.fire("attacker_misses", e.loc, u.loc, {wml.tag.first(w.__cfg),wml.tag.second(w2_cfg)})
```

CelticMinstrel commented on data/lua/intercept_graphics.lua:
```diff
@@ -0,0 +1,63 @@
+local function clear_halo_locations() -- shared cleanup logic
+    if next(halo_locations) ~= nil then --check if halo_locations is empty so that redraw is not unnecessarily called
+        for _, loc in ipairs(halo_locations) do
+            wesnoth.interface.remove_hex_overlay(loc[1], loc[2], halo_id)
```
It would be much preferable if these locations were stored as `{x = xcoord, y = ycoord}` instead of `{x, y}`, though I don't consider this one a blocker since it's entirely private to this file.

CelticMinstrel commented on data/lua/intercept_graphics.lua:
```diff
@@ -0,0 +1,63 @@
+            if not wesnoth.units.get(candidate_hex) then -- not placing under units, those tiles cannot be moved onto and the overlay looks weird placed under units
+                local n, ne, se, s, sw, nw = wesnoth.map.get_adjacent_hexes(candidate_hex)
+                for _, loc in ipairs({n, ne, se, s, sw, nw}) do
+                    local e = wesnoth.units.get(loc[1], loc[2])
```
This could be simplified to:
```suggestion
                    local e = wesnoth.units.get(loc)
```

CelticMinstrel reviewed (changes requested):
I think this mostly looks good now. Other than the CI issue, I believe the last thing to do is using `wml.tag` for constructing WML tables.

hladkyjakub committed:
Refactor intercept ability checks and halo updates

hladkyjakub committed:
Refactor intercept weapon special configuration

hladkyjakub committed:
Update Youth.cfg to modify weapon special

Removed USES_PER_TURN from WEAPON_SPECIAL_INTERCEPT.

hladkyjakub committed:
Refactor intercept logic to use special_type

CelticMinstrel commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+    e.variables.intercept_used = e.variables.intercept_used and e.variables.intercept_used + 1 or 1
+    local attacking_weapon
+    for i, attack in ipairs(u.attacks) do
+        if wml.child_array(attack.specials,"intercept")[1] then
```
Again:
```suggestion
        if wml.get_child(attack.specials,"intercept") then
```

CelticMinstrel commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+        local __, __, e_stats, u_stats = wesnoth.simulate_combat(e,attacking_weapon, u)-- Defensive specials of enemy weapon do apply, but enemy doesn't fight back - makes sence for things like shield block
+        local w = e.attacks[attacking_weapon]
+
+        local strike_limit = wml.child_array(w.specials,"intercept")[1].strike_limit
```
Again:
```suggestion
        local strike_limit = wml.get_child(w.specials,"intercept").strike_limit
```

hladkyjakub commented on data/lua/intercept_graphics.lua:
```diff
@@ -0,0 +1,85 @@
+		clear_halo_locations(x, y)
+		local sx, sy = wesnoth.interface.get_selected_hex()
+		if sx then
+			halos_for_unit(wesnoth.units.get({sx,sy})) --making sure we update after selecting a unit - there are cases, when it's not allready displayed by the time
```
Yes.

hladkyjakub commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+                u_level = u.level
+                u_id= u.id
+                wesnoth.game_events.fire("die", u.loc, e.loc, {{"first",w2_cfg},{"second",w.__cfg}}) --inverted
+                wesnoth.wml_actions.kill({ id = u_id, animate = false, fire_event = false})
```
uhm...
I would think that this cahnge would damage the readability - it's cleaner with the () whenever calling a function, I think.
Are you sure you want me to remove them?

hladkyjakub replied:
Pending.
I still do belive that () cause no harm and improve readability there.

hladkyjakub committed:
Merge branch 'wesnoth:master' into Weapon-Special-Intercept

hladkyjakub committed:
Change attack references from unit to enemy

Updated the attack references from the unit to the enemy in the intercept logic.

hladkyjakub committed:
Refactor intercept ability and halo update logic

Refactor intercept ability functions and update event handling.

hladkyjakub committed:
Add reachmap_updated_callback function declaration

hladkyjakub committed:
Implement reachmap_updated callback in game_lua_kernel

Added a new callback for reachmap updates in the Lua kernel.

hladkyjakub committed:
moved reach_map typedef in display.hpp

hladkyjakub committed:
Integrate Lua callback for reachmap updates

Added Lua callback for reachmap updates.

hladkyjakub committed:
emmylua doc

hladkyjakub commented:
@CelticMinstrel Could you review it? (it's rather short code)

I actually have no idea how units/emmylua works and what language it's written in, I just copied it there the same way similar functions are there documented... the table of locations is probably wrong name

hladkyjakub committed:
Fix indentation in display.cpp

hladkyjakub committed:
Pass radius parameter to old_reachmap_updated function

CelticMinstrel reviewed (commented):
Another quick question. I happened to notice that the `{"tag", "contents"}` named tuple is declared in two different places. Would that have any effect on performance at all?

hladkyjakub commented:
> Another quick question. I happened to notice that the `{"tag", "contents"}` named tuple is declared in two different places. Would that have any effect on performance at all?

Wdym? What table is defined twice? And why would it affect performance?

hladkyjakub commented on data/core/macros/weapon_specials.cfg:
```diff
@@ -256,3 +256,101 @@
+        name= _ "intercept"
+        female_name= _ "intercept"
+        description=_ "When an enemy unit enters an adjacent space, the wielder immediately executes an uncountered full-power intecepting attack with this weapon. The exertion prevents repeated use."
+        description_inactive=_ "Not ready yet. When an enemy unit enters an adjacent space, the wielder immediately executes an uncountered full-power intecepting attack with this weapon. The exertion prevents repeated use."
```
I don't mind it there or not.
So should we have it?

soliton- replied:
Since this pretty much just repeats the normal description it seems fairly useless and just adds work for translators.

stevecotton commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+    end
+    local loc = e.loc
+    local loc2 = u.loc
+    local __, __, e_stats, u_stats = wesnoth.simulate_combat(e,attacking_weapon, u)
```
Just found that assigning to `__` breaks wmlxgettext's tracking of which textdomain is assigned to `_`. That's a bug in wmlxgettext, but forbidding the use of `__` seems good anyway.

If it happens in Lua embedded within a .wml file, the effect is to apparently to inherit the textdomain that was active in the WML part. In a .lua file, the strings aren't recorded at all.

hladkyjakub committed:
Update data/core/macros/weapon_specials.cfg

Co-authored-by: Gunter Labes <soliton.de@gmail.com>

hladkyjakub committed:
Update weapon_specials.cfg

hladkyjakub commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+--attacker is e
+--defender is u
+
+local _ = wesnoth.textdomain "wesnoth-help" -- !! This takes away "_" as commonly used unused variable !!
```
Ok. I wasn't sure where the translations are at.

hladkyjakub committed:
Change textdomain from 'wesnoth-help' to 'wesnoth'

hladkyjakub commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+for iter,e in ipairs(enemies) do
+    local attacking_weapon
+    for i, attack in ipairs(u.attacks) do
+        if wml.child_array(attack.specials,"intercept")[1] then
```
As settled elsewhere - ready for future work.

hladkyjakub commented on data/lua/intercept.lua:
```diff
@@ -0,0 +1,186 @@
+--attacker is e
+--defender is u
```
It's for future work, will simplify that once proper API to handle the attack and its animations are made, or I find a suitable way to pack the code into functions.

Pentarctagon commented:
So, what's the current status of this PR?
